File: Python3/프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    dr, dc = [0, 1, 0, -1], [1, 0, -1, 0]
    answer = [[0] * n for i in range(n)]
    r, c = 0, 0
    d = 0
    nr, nc = 0, 0
    
    for i in range(1, n ** 2 + 1):
        answer[r][c] = i
        nr, nc = r + dr[d], c + dc[d]
        if nr < 0 or n <= nr or nc < 0 or n <= nc or answer[nr][nc] != 0:
            d = (d + 1) % 4
            nr, nc = r + dr[d], c + dc[d]
        r, c = nr, nc
        
    return answer

File: C++/프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> solution(int n) {
    vector<vector<int>> answer;
    for(int i = 0; i < n; i++) answer.push_back(vector<int>(n));
    
    int r = 0, c = 0, nr = 0, nc = 0, d = 0;
    vector<int> dr = {0, 1, 0, -1}, dc = {1, 0, -1, 0};
    for(int i = 1; i <= n * n; i++) {
        answer[r][c] = i;
        nr = r + dr[d]; nc = c + dc[d];
        if(nr < 0 || n <= nr || nc < 0 || n <= nc || answer[nr][nc] != 0) {
            d = (d + 1) % 4;
            nr = r + dr[d]; nc = c + dc[d];
        }
        r = nr; c = nc;
    }
    return answer;
}